docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +180,233 @@ Cluster Service Administration
 +----------+---------------+-------------+----------------------------+
 | edit     | dex-config    | name        | open an editor             |
 +----------+---------------+-------------+----------------------------+
+
+Design Choices
+--------------
+
+Two distinct CLI:
+
+- a ``metalk8s`` kubectl plugin with subcommands to interact with Kubernetes
+  API, and Salt API if needed.
+- a ``metalk8sctl`` CLI with subcommands for action that need to interact with
+  the local machine, but may also interact with Kubernetes API and Salt API if
+  needed.
+
+
+``metalk8s`` kubectl plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Go`` is chosen as the language for kubectl plugin for the following reasons:
+
+- Great interactions with Kubernetes API.
+- Often used for operators and kubectl plugins
+  (`Sample CLI plugin <https://github.com/kubernetes/sample-cli-plugin>`_,
+  `Helpers for kubectl plugins <https://github.com/kubernetes/cli-runtime>`_).
+- Portable as it's just a binary.
```

Comment:
It's "portable" because it's a statically compiled binary, being a binary doesn't make you portable at all per se.
Also, "portable" may not be the right term here (won't run on other OSes or CPU architecture, not our use case anyway).

It's more like "easy to ship because it's a statically compiled binary, no deps to provide".

Reply:
You right, I will update this.

> Also, "portable" may not be the right term here (won't run on other OSes or CPU architecture, not our use case anyway).

Hmm it's one use case here because this kubectl plugin may run from customer laptops so likely any OSes

Reply:
> Hmm it's one use case here because this kubectl plugin may run from customer laptops so likely any OSes

As long as it's x86_64 Linux OS it should be fine (i.e, it's portable across Ubuntu, Fedora, Archlinux, CentOS, …).

But if the customer wants to run it from MacOS, Windows, or some ARM board I'm not sure it will work out of the box.

`Go` can cross-compile and generate binaries for those platform quite easily (as long as our dependencies also works on those systems/architecture), but that will be a distinct binary (example [here](https://github.com/helm/helm/releases), where `helm` provide one binary for each case).

So yeah, it can be truly portable, but that's not "one binary to run them all".

So, given our dependencies itself are portable, our plugin can be portable too but the user may have to compile it ourself (or we need to provide/release cross-compiled binary like `helm` does).

Reply:
Ok good to know then I will changes this one for the moment and we will think about nno-linux x86_64 later (I will add a todo for this) :+1: